s3stream/src/main/java/com/automq/stream/s3/S3Storage.java

```diff
@@ -596,6 +598,32 @@ private void continuousCheck(List<StreamRecordBatch> records) {
         }
     }
 
+    private void maybeForceUpload() {
+        // Limit the number of inflight force upload tasks to avoid too many S3 objects.
+        if (hasInflightForceUploadTask()) {
```

Comment:
The inflightWALUploadTasks and needForceUpload may be accessed by multiple threads